### PR TITLE
Store serialized JSON ACME Account as String

### PR DESCRIPTION
**Pull request by alwinmarkcf**
### What does this PR do?

Before it was stored as []byte which make it hard to backup ACME
Certificates and debug the behaviour of getting new Certificates. So this PR changes this behaviour and stores the serialized JSON ACME Account as String

### Motivation

Basically the motivation were these two issues, as for a productive usage of traefik it's a must have to backup certs. #1512

Also I have several problems during an upgrade process with kubernetes I can't really debug without knowing what happens there: #3754

### More

- [ ] Added/updated tests
- [ ] Added/updated documentation

### Additional Notes

Actually I don't know why this Meta object is used for the Account. Guess using [staert](https://github.com/containous/staert) with the Account struct itself could remove the limitation: 
> !!! note It's possible to store up to approximately 100 ACME certificates in Consul.

Otherwise it would make more sense to have a dedicated storeMethod/Transferobject for the Account or certificate.

Described here: https://github.com/nmengin/traefik/blob/357d8762a16405a881d72907c39f61d78372a26f/docs/configuration/acme.md
Issued here: https://github.com/containous/traefik/issues/1325

Also this may bring up again the topic about: https://github.com/containous/traefik/issues/1325 but where #1325 makes it hard having multiple certs, the current behaviour makes it impossible to backup certs which prevents the usage of traefik in production

### CODE REVIEW

**Comment by alwinmarkcf** [comment]
I don't know how the process is for merging into multiple branches.

**Comment by atombender** [comment]
Doesn't this need to have a fallback to the old byte format so that it can read the existing data that people currently have stored in their clusters?

**Comment by emilevauge** [comment]
Hey @alwinmarkcf 
Thanks for your contribution!

There are 2 major regressions brought by this fix:
- this is not backward compatible with the current implementation
- more importantly, this removes the compression we added https://github.com/containous/traefik/pull/2814, which allows to store up to approximately 100 ACME certificates

The second one is probably a deal breaker here, as it will not be possible to get the best of both worlds (store many certificates AND human readable format).

In order to help users debug things, we could instead add a new command to Traefik, something like `traefik debugConfig`, that would read the kv store, and print the content on std out. WDYT?

**Comment by atombender** [comment]
Could there be (or is there) a flag to toggle compression? Etcd has no such limit on the size of the value, so the binary format is of no use there.

**Comment by atombender** [comment]
Also, why not just gzip the JSON? The curren object appears to be a binary structure. If it were gzipped JSON, you could just do `etcdctl get --print-value-only /traefik/acme/account/object | gzip -dc` to read it.

**Comment by alwinmarkcf** [comment]
@emilevauge so in general not specific to this PR I have following usecases, which are in my opinion needed to make Traefik really "cloudnative" and production ready.

Unfortunately I kind of love this project and the idea behind it and we as crowdfox put some effort into it, so it would be awesome if we could help bringing traefik into this direction.

_Actually I don't know where to post this without loosing the overview as its more on an epic level then an issue. Maybe you can move it to the right place._

## Helm install

```gherkin
Given a proper values.yaml for HA mode
When I do a "helm install"
Then  traefik should run in HA mode
```
This one works since we got: https://github.com/helm/charts/pull/6813 merged

## Upgrade

### simple upgrade

```gherkin
Given a proper values.yaml for HA mode
And traefik is installed via helm in HA mode
When I change some parameter in the values.yaml
Then I these changes should take effect
```

For this one, I'd love to see what's discussed here: https://github.com/containous/traefik/issues/3754
Tl;dr:
* Keep configs in TOML and only store states other than configs in KV Store (Like ACME Certs). 
* make Traefik start with empty kv set and generate them on the fly

### Idempotent upgrade

```gherkin
Given a proper values.yaml for HA mode
And traefik is installed via helm in HA mode
When I change some parameter in the values.yaml
And I install on a different cluster traefik with the same values.yaml from scratch
Then both traefik server should behave the same
```

Would be fixed when only the config file would be considered. The Problem here currently is that storeconfig does not clean Keys, so you have legacy key/values which you only can clean manually. Thats a huge Problem. Eg: `traefik/acme/domains`

## Backup

### create backup

```gherkin
Given traefik is installed 
When I run a backup command
Then I want to be able to backup all states which are not given by:
* Providers
* Configuration config.tom
(means especially ACME certificates)
```
This would work merging this Ticket, even though it would strongly reduce max number of certs for consul. But then you could backup your config running:
```bash
kubectl run  --rm -it traefik-etcd-test --image=quay.io/coreos/etcd:v3.1.8 --restart=Never --env="ETCDCTL_API=3" -l app=traefik -- etcdctl --endpoints=http://traefik-etcd-cluster-client:2379 get traefik/acme/domains --prefix --print-value-only | jq .
```

### import backup

```gherkin
Given traefik is installed
When I run an import backup command (e.g. storeconfig)
Then all states given by this backup should be restored (especially ACME certs)
```
This is some kind of working now, using the storeconfig job, https://github.com/helm/charts/pull/7173, and paramter: `--set traefik.kvprovider.importAcme=true`

## ACME Troubleshooting

### Refetch certificate

Given an invalid/lost cert
When I run update cert for domain "example.dev"
Then traefik should try to get a new cert for this domain

While having the ability to export the config and import it back you could do this also manually.

### Healthcheck

https://github.com/containous/traefik/issues/3760

This would be really nice to have a check running if certs are too old and can't be refetched because of ratelimits or a broken letsencrypt client (key)

**Comment by nmengin** [comment]
Hello @atombender,

> Also, why not just gzip the JSON? 

Træfik has used the library `staert` to insert data into the KV stores.
`staert` stores the strings in the KV stores without modifying the data, it only can compress the `[]byte` and modifying this behavior can be hard thats why the JSON is compressed as a `[]byte` and not a `string`.

**Comment by nmengin** [comment]
Hello @alwinmarkcf,

Many thanks for the PR and all the propositions you suggested to improve Træfik.
I guess, it should be better to open an issue per propositions, thus we can analyse and prioritize them.

Concerning the PR, as @emilevauge said previously, the modifications you propose require more work to be backward compatible (like supporting a third format for the ACME Account Key).

That's why I just opened a new issue #3847 to create a subcommand to export the data you need without changing nothing in Træfik.
A PR will come to implement it.

Once again, thanks for your work and all the propositions you have made in this PR and the others issues.

Closed in favor of #3847.
